x/ibc/07-tendermint/update.go

```diff
@@ -0,0 +1,69 @@
+package tendermint
+
+import (
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clientexported "github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	commitment "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment"
+)
+
+// CheckValidityAndUpdateState checks if the provided header is valid and updates
+// the consensus state if appropriate. It returns an error if:
+// - the client or header provided are not parseable to tendermint types
+// - the header is invalid
+// - header height is lower than the latest client height
+// - header valset commit verification fails
+func CheckValidityAndUpdateState(
+	clientState clientexported.ClientState, header clientexported.Header, chainID string,
+) (clientexported.ClientState, clientexported.ConsensusState, error) {
+	tmClientState, ok := clientState.(ClientState)
+	if !ok {
+		return nil, nil, sdkerrors.Wrap(
+			clienttypes.ErrInvalidClientType, "light client is not from Tendermint",
+		)
+	}
+
+	tmHeader, ok := header.(Header)
+	if !ok {
+		return nil, nil, sdkerrors.Wrap(
+			clienttypes.ErrInvalidHeader, "header is not from Tendermint",
+		)
+	}
+
+	if err := checkValidity(tmClientState, tmHeader, chainID); err != nil {
+		return nil, nil, err
+	}
+
+	tmClientState, consensusState := update(tmClientState, tmHeader)
+	return tmClientState, consensusState, nil
+}
+
+// checkValidity checks if the Tendermint header is valid
+//
+// CONTRACT: assumes header.Height > consensusState.Height
+func checkValidity(clientState ClientState, header Header, chainID string) error {
+	if header.GetHeight() < clientState.LatestHeight {
+		return sdkerrors.Wrapf(
+			clienttypes.ErrInvalidHeader,
+			"header height < latest client state height (%d < %d)", header.GetHeight(), clientState.LatestHeight,
+		)
+	}
+
+	// basic consistency check
+	if err := header.ValidateBasic(chainID); err != nil {
+		return err
+	}
+
```

Comment:
Seems like with removing the `Committer` interface an important check was removed.

I believe we still need to make sure that this Header will be verified by the last update's ValidatorSet.

To do this, we need a way to retrieve the last ValidatorSet that caused an update on this connection and then call

`lastValidatorSet.VerifyFutureCommit(header.ChainID, header.Commit.BlockID, header.Height, header.Commit)`

Not sure how to do this with the current datastructures now that Committer is removed. If the full ValidatorSet gets saved in `ConsensusState` rather than just the hash then this is possible

The code as is seems like it will allow arbitrary updates to a connection from an arbitrary validatorset.

Need to implement the bisection algorithm for correct and safe updating

Reply:
> If the full ValidatorSet gets saved in ConsensusState rather than just the hash then this is possible

@cwgoes can we update the spec to include the validator set instead of the hash?

Reply:
Sure. https://github.com/cosmos/ics/issues/366. I'll work on this now.